### PR TITLE
feat: suspend/resume the vm

### PR DESCRIPTION
**Pull request by mohanson**
For #109

### CODE REVIEW

File: src/machine/aot/aot.x64.c [hunk 1]
```diff
@@ -369,6 +369,14 @@ int aot_link(AotContext* context, size_t *szp)
   | and edx, CKB_VM_ASM_MEMORY_FLAG_WXORX_BIT
   | cmp edx, CKB_VM_ASM_MEMORY_FLAG_WRITABLE
   | jne >4
+  /*
+   * Set the page as dirty
+   */
+  | lea rdx, machine->flags
+  | movzx edx, byte [rdx+rcx]
+  | and edx, CKB_VM_ASM_MEMORY_FLAG_DIRTY
```
Comment (xxuejie) [inline src/machine/aot/aot.x64.c]: I think this should be an or, correct?

File: src/machine/aot/aot.x64.c [hunk 2]
```diff
@@ -369,6 +369,14 @@ int aot_link(AotContext* context, size_t *szp)
...
+  | lea rdx, machine->flags
```
Comment (xxuejie) [inline src/machine/aot/aot.x64.c]: This line already changes the value of rdx, which means `dl` below is not a correct value.

File: src/snapshot.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    snap.pc = machine.pc().to_u64();
+
+    for i in 0..RISCV_PAGES {
+        let flag = machine.memory_mut().fetch_flag(i as u64)?;
+        if flag & FLAG_DIRTY != 0 {
+            let addr_from = i << RISCV_PAGE_SHIFTS;
+            let addr_to = (i + 1) << RISCV_PAGE_SHIFTS;
+
+            let mut page = vec![0; RISCV_PAGESIZE];
+            for i in (addr_from..addr_to).step_by(8) {
```
Comment (xxuejie) [inline src/snapshot.rs:54]: Can we add some comments to document how snapshot works?

File: src/machine/aot/aot.x64.c (hunk 2 above)
Comment (mohanson) [reply src/machine/aot/aot.x64.c]: Yes, I wrote the wrong logic.

File: src/snapshot.rs (hunk 3 above)
Comment (mohanson) [reply src/snapshot.rs:54]: I submitted an update, and now the comments will explain what we did.

```rs
// Snapshot provides a mechanism for suspending and resuming a virtual machine.
//
// When cycle limit is too low, our work won't be finished when we hit it and
// a "max cycles exceeded" error will be returned. At this time, a snapshot could
// be created, we can create a new virtual machine from this snapshot, provide
// more cycles and then continue to run it.
//
// For the following data, we simply save them, and then restore them.
//   - machine.version
//   - machine.pc
//   - machine.registers
//
// For memory, the situation becomes more complicated. Every memory page has
// page flag where each page flag stores a optional FLAG_DIRTY. When this page
// is wrote, the memory instance will set this bit to 1, this helps us keep
// track of those pages that have been modified. After `machine.load_elf`, We
// clean up all dirty flags, so after the program terminates, all pages marked
// as dirty are the pages that have been modified by the program. We only store
// these pages in the snapshot.
```

**Review by xxuejie — APPROVED** [review]

**Pull request merged**